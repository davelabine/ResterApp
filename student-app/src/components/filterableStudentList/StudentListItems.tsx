import * as React from 'react';

import { StudentListItem } from './StudentListItem';
import { StudentData } from './StudentListInterfaces.d';

export interface StudentListItemsProps {
    filter: string;
    students: Array<StudentData>;
}

/* TODO: There is probably a better way to do this without polluting the global namespace.
   Initial attempt at adding this as a class constant didn't work - could not access from another class */
export const STUDENT_LIST_ITEMS_EMPTY: string = 'No Students!';

export class StudentListItems extends React.Component<StudentListItemsProps, object> {
    
    constructor(props: StudentListItemsProps) {
        super(props);
    }

    public render() {
        const studentList = this.filterRows();

        return (
            <table className="studentListItems table table-striped table-sm table-hover">
                <thead className="studentListItemsHead">
                <tr>
                    <th>Name</th>
                    <th>ID</th>
                    <th>URL</th>
                </tr>
                </thead>
                <tbody className="studentListItemsBody">
                    {studentList}
                </tbody>
            </table>
        );
    }

    private filterRows(): Array<JSX.Element> {
        const filter = this.props.filter;
        const students = this.props.students;
        const rows: Array<JSX.Element> = [];

        students.map((s) => {
            if (s.name.toLowerCase().indexOf(filter.toLowerCase()) === -1) {
                return;
            }

            rows.push(
                <StudentListItem key={s.skey} student={s}/>
            );
        });

        /* Note: Empty rows could be because of no students in the list
           or we filtered all of the existing students out */ 
        if (!rows.length) {
            rows.push(<tr key="empty"><td>{STUDENT_LIST_ITEMS_EMPTY}</td></tr>);
        }

        return rows;
    }
}
export default StudentListItems;