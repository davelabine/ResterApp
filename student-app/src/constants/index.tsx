export const FETCH_STUDENTS = 'FETCH_STUDENTS';
export type FETCH_STUDENTS = typeof FETCH_STUDENTS;

export const FILTER_STUDENTS = 'FILTER_STUDENTS';
export type FILTER_STUDENTS = typeof FILTER_STUDENTS;

export const ADD_STUDENT = 'ADD_STUDENT';
export type ADD_STUDENT = typeof ADD_STUDENT;