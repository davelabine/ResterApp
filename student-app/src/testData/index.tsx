import { StudentData } from '../types';

export const STUDENT_DATA_EMPTY: StudentData = new StudentData();

export const STUDENT_DATA_FAKE: StudentData =  {
    'skey': 'FAKE_KEY',
    'id': 'FAKE_ID',
    'lastName': 'FAKE_NAME',
    'photo': {}
};

export const STUDENT_DATA_BILLY: StudentData =  {
                                                    'skey': '8a80810d5dd89a3f015dd89ac8280000',
                                                    'id': '1234',
                                                    'lastName': 'Billy Bob',
                                                    'photo': {
                                                        'bucketName': 'resterapp-dev',
                                                        'key': 'e93ecbf0-ba96-4290-a906-5ea67e4c4a9f'
                                                    }
                                                };

export const STUDENT_DATA_BILLY_UPDATE: StudentData =  {
                                                    'skey': '8a80810d5dd89a3f015dd89ac8280000',
                                                    'id': '007',
                                                    'lastName': 'James Bond',
                                                    'photo': {}
                                                };

export const LIST_STUDENT_DATA_EMPTY: StudentData[] = [];

export const LIST_STUDENT_DATA_TWO: StudentData[] =  [
    {
        'skey': '8a80810d5dd89a3f015dd89ac8280000',
        'id': '1234',
        'lastName': 'Billy Bob',
        'photo': {
            'bucketName': 'resterapp-dev',
            'key': 'e93ecbf0-ba96-4290-a906-5ea67e4c4a9f'
        }
    },
    {
        'skey': '8a80810d5dd89a3f015dd8a252800001',
        'id': '503074',
        'lastName': 'Albert Russell',
        'photo': {
            'bucketName': 'resterapp-dev',
            'key': 'e93ecbf0-ba96-4290-a906-5ea67e4c4a9f'
        }
    },
];

export const LIST_STUDENT_DATA_BIG: StudentData[] =  [
    {
        'skey': '8a80810d5dd89a3f015dd89ac8280000',
        'id': '1234',
        'lastName': 'Billy Bob',
        'photo': {
            'bucketName': 'resterapp-dev',
            'key': 'e93ecbf0-ba96-4290-a906-5ea67e4c4a9f'
        }
    },
    {
        'skey': '8a80810d5dd89a3f015dd8a252800001',
        'id': '503074',
        'lastName': 'Albert Russell',
        'photo': {
            'bucketName': 'resterapp-dev',
            'key': 'e93ecbf0-ba96-4290-a906-5ea67e4c4a9f'
        }
    },
    {
        'skey': '8a80810d5dd89a3f015dd8a259b2000e',
        'id': '942226',
        'lastName': 'Alfredo Cummings',
        'photo': {}
    },
    {
        'skey': '8a80810d5dd89a3f015dd8a2677f0027',
        'id': '824388',
        'lastName': 'Allen Russell',
        'photo': {}
    },
    {
        'skey': '8a80810d5dd89a3f015dd8a271fb003a',
        'id': '266192',
        'lastName': 'Andre McKenna',
        'photo': {}
    },
    {
        'skey': '8a80810d5dd89a3f015dd8a27f0d0052',
        'id': '932083',
        'lastName': 'Angelica Obrien',
        'photo': {}
    },
    {
        'skey': '8a80810d5dd89a3f015dd8a281c80057',
        'id': '682307',
        'lastName': 'Anna Washington',
        'photo': {}
    },
    {
        'skey': '8a80810d5dd89a3f015dd8a27c52004d',
        'id': '734552',
        'lastName': 'Antonio Fowler',
        'photo': {}
    },
    {
        'skey': '8a80810d5dd89a3f015dd8a25fb80019',
        'id': '615643',
        'lastName': 'Ashley Mcdaniel',
        'photo': {}
    },
    {
        'skey': '8a80810d5dd89a3f015dd8a25897000c',
        'id': '753238',
        'lastName': 'Beatrice Russell',
        'photo': {}
    },
];
