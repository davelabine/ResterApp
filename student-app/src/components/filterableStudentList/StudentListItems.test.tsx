import * as React from 'react';
import { shallow, mount } from 'enzyme';
import { StudentListItems } from './StudentListItems';
import * as constants from '../../constants/index';
import * as studentTestData from '../../testData';
import toJson from 'enzyme-to-json';

describe('StudentListItems', () => {

  it('renders without crashing and matches the last snapshot', () => {
    const list = mount(<StudentListItems filter="" students={studentTestData.LIST_STUDENT_DATA_TWO}/>);
    expect(toJson(list)).toMatchSnapshot();
  }); 

  it('renders the right number of studentListItems', () => {
    const list = shallow(<StudentListItems filter="" students={studentTestData.LIST_STUDENT_DATA_TWO}/>);
    const listRows = list.find('.studentListItemsBody').children();
    expect(listRows.length).toEqual(2);
  });

  it('renders a message when it gets an empty student list', () => {
    const list = shallow(<StudentListItems filter="" students={studentTestData.LIST_STUDENT_DATA_EMPTY}/>);
    const listRows = list.find('.studentListItemsBody').children();
    expect(listRows.length).toEqual(1);
    expect(listRows.first().text()).toEqual(constants.STUDENT_LIST_ITEMS_EMPTY);
  });

  it('filters out all but the A students from a two student list', () => {
    const list = shallow(<StudentListItems filter="A" students={studentTestData.LIST_STUDENT_DATA_TWO}/>);
    const listRows = list.find('.studentListItemsBody').children();
    expect(listRows.length).toEqual(1);
  });

  it('ignores case when filtering students from a two student list', () => {
    const list = shallow(<StudentListItems filter="A" students={studentTestData.LIST_STUDENT_DATA_TWO}/>);
    const listRows = list.find('.studentListItemsBody').children();
    expect(listRows.length).toEqual(1);
  });

  it('properly filters out all students from a two student list and displays the empty student message', () => {
    const list = shallow(<StudentListItems filter="Z" students={studentTestData.LIST_STUDENT_DATA_TWO}/>);
    const listRows = list.find('.studentListItemsBody').children();
    expect(listRows.length).toEqual(1);
    expect(listRows.first().text()).toEqual(constants.STUDENT_LIST_ITEMS_EMPTY);
  });
});