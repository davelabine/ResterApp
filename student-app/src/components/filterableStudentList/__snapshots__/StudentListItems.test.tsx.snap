// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StudentListItems renders without crashing and matches the last snapshot 1`] = `
<StudentListItems
  filter=""
  onDeleteStudent={[Function]}
  onUpdateStudent={[Function]}
  students={
    Array [
      Object {
        "id": "1234",
        "lastName": "Billy Bob",
        "photo": Object {
          "bucketName": "resterapp-dev",
          "key": "e93ecbf0-ba96-4290-a906-5ea67e4c4a9f",
        },
        "skey": "8a80810d5dd89a3f015dd89ac8280000",
      },
      Object {
        "id": "503074",
        "lastName": "Albert Russell",
        "photo": Object {
          "bucketName": "resterapp-dev",
          "key": "e93ecbf0-ba96-4290-a906-5ea67e4c4a9f",
        },
        "skey": "8a80810d5dd89a3f015dd8a252800001",
      },
    ]
  }
>
  <div>
    <Table
      bordered={false}
      bsClass="table"
      condensed={true}
      hover={true}
      responsive={false}
      striped={true}
    >
      <table
        className="table table-striped table-condensed table-hover"
      >
        <thead
          className="studentListItemsHead"
        >
          <tr>
            <th>
              Name
            </th>
            <th>
              ID
            </th>
            <th>
              URL
            </th>
          </tr>
        </thead>
        <tbody
          className="studentListItemsBody"
        >
          <StudentListItem
            key="8a80810d5dd89a3f015dd89ac8280000"
            onDeleteStudent={[Function]}
            onUpdateStudent={[Function]}
            student={
              Object {
                "id": "1234",
                "lastName": "Billy Bob",
                "photo": Object {
                  "bucketName": "resterapp-dev",
                  "key": "e93ecbf0-ba96-4290-a906-5ea67e4c4a9f",
                },
                "skey": "8a80810d5dd89a3f015dd89ac8280000",
              }
            }
          >
            <tr>
              <td>
                Billy Bob
              </td>
              <td>
                1234
              </td>
              <td>
                <ButtonGroup
                  block={false}
                  bsClass="btn-group"
                  justified={false}
                  vertical={false}
                >
                  <div
                    className="btn-group"
                  >
                    <Button
                      active={false}
                      block={false}
                      bsClass="btn"
                      bsStyle="default"
                      disabled={false}
                      id="edit"
                      onClick={[Function]}
                    >
                      <button
                        className="btn btn-default"
                        disabled={false}
                        id="edit"
                        onClick={[Function]}
                        type="button"
                      >
                        <Glyphicon
                          bsClass="glyphicon"
                          glyph="pencil"
                        >
                          <span
                            className="glyphicon glyphicon-pencil"
                          />
                        </Glyphicon>
                      </button>
                    </Button>
                    <Button
                      active={false}
                      block={false}
                      bsClass="btn"
                      bsStyle="default"
                      disabled={false}
                      id="delete"
                      onClick={[Function]}
                    >
                      <button
                        className="btn btn-default"
                        disabled={false}
                        id="delete"
                        onClick={[Function]}
                        type="button"
                      >
                        <Glyphicon
                          bsClass="glyphicon"
                          glyph="trash"
                        >
                          <span
                            className="glyphicon glyphicon-trash"
                          />
                        </Glyphicon>
                      </button>
                    </Button>
                  </div>
                </ButtonGroup>
                <EditStudentModal
                  initialStudent={
                    Object {
                      "id": "1234",
                      "lastName": "Billy Bob",
                      "photo": Object {
                        "bucketName": "resterapp-dev",
                        "key": "e93ecbf0-ba96-4290-a906-5ea67e4c4a9f",
                      },
                      "skey": "8a80810d5dd89a3f015dd89ac8280000",
                    }
                  }
                  onHide={[Function]}
                  onSubmit={[Function]}
                  show={false}
                  submitButtonText="Save"
                  title="Edit Billy Bob"
                >
                  <div
                    className="static-modal"
                  >
                    <Modal
                      animation={true}
                      autoFocus={true}
                      backdrop={true}
                      bsClass="modal"
                      dialogClassName="custom-modal"
                      dialogComponentClass={[Function]}
                      enforceFocus={true}
                      keyboard={true}
                      manager={
                        ModalManager {
                          "add": [Function],
                          "containers": Array [],
                          "data": Array [],
                          "handleContainerOverflow": true,
                          "hideSiblingNodes": true,
                          "isTopModal": [Function],
                          "modals": Array [],
                          "remove": [Function],
                        }
                      }
                      onHide={[Function]}
                      renderBackdrop={[Function]}
                      restoreFocus={true}
                      show={false}
                    >
                      <Modal
                        autoFocus={true}
                        backdrop={true}
                        backdropClassName="modal-backdrop"
                        backdropTransitionTimeout={150}
                        containerClassName="modal-open"
                        dialogTransitionTimeout={300}
                        enforceFocus={true}
                        keyboard={true}
                        manager={
                          ModalManager {
                            "add": [Function],
                            "containers": Array [],
                            "data": Array [],
                            "handleContainerOverflow": true,
                            "hideSiblingNodes": true,
                            "isTopModal": [Function],
                            "modals": Array [],
                            "remove": [Function],
                          }
                        }
                        onEntering={[Function]}
                        onExited={[Function]}
                        onHide={[Function]}
                        renderBackdrop={[Function]}
                        restoreFocus={true}
                        show={false}
                        transition={[Function]}
                      />
                    </Modal>
                  </div>
                </EditStudentModal>
              </td>
            </tr>
          </StudentListItem>
          <StudentListItem
            key="8a80810d5dd89a3f015dd8a252800001"
            onDeleteStudent={[Function]}
            onUpdateStudent={[Function]}
            student={
              Object {
                "id": "503074",
                "lastName": "Albert Russell",
                "photo": Object {
                  "bucketName": "resterapp-dev",
                  "key": "e93ecbf0-ba96-4290-a906-5ea67e4c4a9f",
                },
                "skey": "8a80810d5dd89a3f015dd8a252800001",
              }
            }
          >
            <tr>
              <td>
                Albert Russell
              </td>
              <td>
                503074
              </td>
              <td>
                <ButtonGroup
                  block={false}
                  bsClass="btn-group"
                  justified={false}
                  vertical={false}
                >
                  <div
                    className="btn-group"
                  >
                    <Button
                      active={false}
                      block={false}
                      bsClass="btn"
                      bsStyle="default"
                      disabled={false}
                      id="edit"
                      onClick={[Function]}
                    >
                      <button
                        className="btn btn-default"
                        disabled={false}
                        id="edit"
                        onClick={[Function]}
                        type="button"
                      >
                        <Glyphicon
                          bsClass="glyphicon"
                          glyph="pencil"
                        >
                          <span
                            className="glyphicon glyphicon-pencil"
                          />
                        </Glyphicon>
                      </button>
                    </Button>
                    <Button
                      active={false}
                      block={false}
                      bsClass="btn"
                      bsStyle="default"
                      disabled={false}
                      id="delete"
                      onClick={[Function]}
                    >
                      <button
                        className="btn btn-default"
                        disabled={false}
                        id="delete"
                        onClick={[Function]}
                        type="button"
                      >
                        <Glyphicon
                          bsClass="glyphicon"
                          glyph="trash"
                        >
                          <span
                            className="glyphicon glyphicon-trash"
                          />
                        </Glyphicon>
                      </button>
                    </Button>
                  </div>
                </ButtonGroup>
                <EditStudentModal
                  initialStudent={
                    Object {
                      "id": "503074",
                      "lastName": "Albert Russell",
                      "photo": Object {
                        "bucketName": "resterapp-dev",
                        "key": "e93ecbf0-ba96-4290-a906-5ea67e4c4a9f",
                      },
                      "skey": "8a80810d5dd89a3f015dd8a252800001",
                    }
                  }
                  onHide={[Function]}
                  onSubmit={[Function]}
                  show={false}
                  submitButtonText="Save"
                  title="Edit Albert Russell"
                >
                  <div
                    className="static-modal"
                  >
                    <Modal
                      animation={true}
                      autoFocus={true}
                      backdrop={true}
                      bsClass="modal"
                      dialogClassName="custom-modal"
                      dialogComponentClass={[Function]}
                      enforceFocus={true}
                      keyboard={true}
                      manager={
                        ModalManager {
                          "add": [Function],
                          "containers": Array [],
                          "data": Array [],
                          "handleContainerOverflow": true,
                          "hideSiblingNodes": true,
                          "isTopModal": [Function],
                          "modals": Array [],
                          "remove": [Function],
                        }
                      }
                      onHide={[Function]}
                      renderBackdrop={[Function]}
                      restoreFocus={true}
                      show={false}
                    >
                      <Modal
                        autoFocus={true}
                        backdrop={true}
                        backdropClassName="modal-backdrop"
                        backdropTransitionTimeout={150}
                        containerClassName="modal-open"
                        dialogTransitionTimeout={300}
                        enforceFocus={true}
                        keyboard={true}
                        manager={
                          ModalManager {
                            "add": [Function],
                            "containers": Array [],
                            "data": Array [],
                            "handleContainerOverflow": true,
                            "hideSiblingNodes": true,
                            "isTopModal": [Function],
                            "modals": Array [],
                            "remove": [Function],
                          }
                        }
                        onEntering={[Function]}
                        onExited={[Function]}
                        onHide={[Function]}
                        renderBackdrop={[Function]}
                        restoreFocus={true}
                        show={false}
                        transition={[Function]}
                      />
                    </Modal>
                  </div>
                </EditStudentModal>
              </td>
            </tr>
          </StudentListItem>
        </tbody>
      </table>
    </Table>
  </div>
</StudentListItems>
`;
